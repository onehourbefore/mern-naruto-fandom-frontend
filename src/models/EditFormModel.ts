

export enum TasksEnum {
    CREATE= 'create',
    UPDATE = 'update',
    IDLE = 'idle'
}

export interface EditFormState {
    postID: string
    task: TasksEnum
}