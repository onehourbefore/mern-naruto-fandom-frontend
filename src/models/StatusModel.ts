export enum StatusModel {
    LOADING = 'loading',
    ERROR = 'error',
    SUCCESS = 'success',
    IDLE = 'idle'
}