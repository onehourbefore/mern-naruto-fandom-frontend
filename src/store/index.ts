import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux';

import posts from './slices/postsSlice';
import filter from './slices/filterSlice';
import onePost from './slices/onePostSlice';
import registration from './slices/registrationSlice';
import authorization from './slices/authorizationSlice';
import experts from './slices/expertSlice';
import comments from './slices/commentSlice';
import editForm from './slices/editFormSlice';

export const rootReducer = combineReducers ({
    posts,
    filter,
    onePost,
    registration,
    authorization,
    experts,
    comments,
    editForm,
});

export const store = configureStore ({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware (),
    devTools: process.env.NODE_ENV !== 'production'
});

export type RootState = ReturnType <typeof store.getState>;
export const useAppDispatch: () => typeof store.dispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;