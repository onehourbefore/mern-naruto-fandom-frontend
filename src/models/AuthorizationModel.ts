import { StatusModel } from "./StatusModel"

export type UserDataType = {
    id: string,
    email: string,
    name: string,
    role: string,
    avatar: string,
    created: number,
    liked: string [] | []
}

export type ResponseType = {
    accessToken: string,
    refreshToken: string,
    user: UserDataType
}

export interface AuthorizationState {
    user: UserDataType,
    status: StatusModel,
    deleteProfileStatus: StatusModel
}