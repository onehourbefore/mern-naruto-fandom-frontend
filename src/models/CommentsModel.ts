import { StatusModel } from "./StatusModel"

export type CommentType = {
    _id: string,
    postID: string,
    date: string,
    name: string,
    email: string,
    body: string,
    avatar: string,
}

export interface CommentsState {
    comments: [] | CommentType [],
    status: StatusModel,
}

export type CommentProps = {
    data: {
        _id: string,
        name: string,
        email: string,
        body: string,
        avatar: string,
        date: string,
    },
    postID: string | undefined,
    commentIndex: number,
}

export type CommentFormProps = {
    setVisible: (arg: boolean) => void,
}

export type CommentFromSidebarProps = {
    comment: CommentType,
}