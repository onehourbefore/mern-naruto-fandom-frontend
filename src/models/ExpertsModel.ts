import { StatusModel } from "./StatusModel"

export type ExpertType = {
    avatar: string,
    email: string,
    name: string,
    created: number
}


export interface ExpertsState {
    experts: [] | ExpertType [],
    status: StatusModel
}