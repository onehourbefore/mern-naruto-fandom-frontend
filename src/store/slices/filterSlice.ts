import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import $api from '../../http';
import { PostType } from '../../models/PostModel';
import { FilterState } from '../../models/FilterModel';
import { StatusModel } from '../../models/StatusModel';
import { FilterThunks } from './thunks/FilterThunks';

const filterThunks = new FilterThunks();

const initialState: FilterState = {
    list: [],
    tags: [],
    fetchStatus: StatusModel.IDLE,
    searchStatus: StatusModel.IDLE,
}

export const findPosts = createAsyncThunk(
    'searchPost',
    async(q: {type: '_query' | '_tag', payload: string}, { rejectWithValue }) => {
        try {
            return await filterThunks.findPostsCallback(q);
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
)

export const getAllTags = createAsyncThunk(
    'getAllTags',
    async(_, { rejectWithValue }) => {
        try {
            return await filterThunks.getAllTagsCallback();
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
)

export const filterSlice = createSlice({
    name: 'filter',
    initialState,
    reducers: {
        clearFiltered: (state) => {
            state.list = [];
        },
        clearAllTags: (state) => {
            state.tags = [];
        },
        clearSearchStatus: (state) => {
            state.searchStatus = StatusModel.IDLE;
        }
    },
    extraReducers: builder => {
        builder
        .addCase(findPosts.pending, (state) => {
            state.searchStatus = StatusModel.LOADING;
        })
        .addCase(findPosts.fulfilled, (state, action: PayloadAction <PostType[]>) => {
            if (action.payload.length !== 0) {
                state.list = action.payload;
                state.searchStatus = StatusModel.SUCCESS;
            } else {
                state.searchStatus = StatusModel.ERROR;
            }
        })
        .addCase(findPosts.rejected, (state) => {
            state.searchStatus = StatusModel.ERROR;
        })


        .addCase(getAllTags.pending, (state) => {
            state.fetchStatus = StatusModel.LOADING;
        })
        .addCase(getAllTags.fulfilled, (state, action: PayloadAction <string []>) => {
            state.tags = action.payload;
            state.fetchStatus = StatusModel.SUCCESS;
        })
        .addCase(getAllTags.rejected, (state) => {
            state.fetchStatus = StatusModel.ERROR;
        })

        .addDefaultCase(() => {})
    }
})

const { actions, reducer } = filterSlice;
export default reducer;
export const { clearFiltered, clearAllTags, clearSearchStatus } = actions;