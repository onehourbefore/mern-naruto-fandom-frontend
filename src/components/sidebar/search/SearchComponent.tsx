import React from 'react';
import { useAppDispatch, useAppSelector } from '../../../store/index';
import { getFilter } from '../../../selectors';
import { findPosts, clearSearchStatus } from '../../../store/slices/filterSlice';
import searchIcon from '../../../assets/png/search.png';
import clearIcon from '../../../assets/png/clear.png';

import cl from './SearchComponent.module.scss';


const SearchComponent: React.FC = (): JSX.Element => {
    const dispatch = useAppDispatch();
    const { searchStatus } = useAppSelector(getFilter);
    const [searchValue, setSearchValue] = React.useState<string>('');
    const debouncedTimer = React.useRef<NodeJS.Timeout>();

    function sendSearchQuery(e: React.ChangeEvent<HTMLInputElement>): void {
        setSearchValue(e.target.value);
        if (debouncedTimer.current) {
            clearTimeout(debouncedTimer.current);
        }
        debouncedTimer.current = setTimeout(() => {
            dispatch(findPosts({
                type: '_query', 
                payload: e.target.value.toUpperCase(),
            }));
            setTimeout(() => {
                dispatch(clearSearchStatus());
            }, 3000);
        }, 1000)
    };

    return (
        <div className={cl.root}>
            <div className={cl.root__search}>
                <div className={cl.root__searchWrapper}>
                    <img 
                        className={cl.root__search_icon}
                        src={searchIcon} 
                        alt="Поиск" />
                    <input 
                        className={cl.root__search_input} 
                        id="searchInput" 
                        type="text" 
                        placeholder="Поиск по блогу..."
                        value={searchValue}
                        onChange={sendSearchQuery}
                    />
                </div>

                {searchValue && 
                    <button
                        data-testid="clearBtn-test"
                        className={cl.root__search_btn}
                        onClick={() => setSearchValue('')}
                    >
                        <img 
                            className={cl.root__search_btn_img}
                            src={clearIcon} 
                            alt="Очистить" />
                    </button>}
            </div>

            {searchStatus === "error" &&
                <div className={cl.root__notFoundStatus}>
                    Ничего не найдено
                </div>}
        </div>
    )
};

export default SearchComponent;