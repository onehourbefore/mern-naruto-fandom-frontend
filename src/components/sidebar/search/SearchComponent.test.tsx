import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as reduxHooks from '../../../store';
import * as actions from '../../../store/slices/filterSlice';
import { StatusModel } from '../../../models/StatusModel';
import SearchComponent from './SearchComponent';


const mockedUseAppDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockedUseAppSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const mockedFindPostsAction = jest.spyOn(actions, 'findPosts');
const mockedClearSearchStatusAction = jest.spyOn(actions, 'clearSearchStatus');



describe('SearchComponent', () => {
    it('basic behavior', () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue({ searchStatus: StatusModel.IDLE });
        const component = render(
            <SearchComponent />
        );
        const searchInput = screen.getByPlaceholderText(/поиск по блогу.../i);

        expect(screen.getByAltText(/поиск/i)).toBeInTheDocument();
        expect(searchInput).toBeInTheDocument();
        expect(screen.queryByTestId('clearBtn-test')).toBeNull();

        userEvent.type(searchInput, 'naruto');
        expect(searchInput).toHaveDisplayValue('naruto');
        expect(screen.getByTestId('clearBtn-test')).toBeInTheDocument();

        userEvent.click(screen.getByTestId('clearBtn-test'));
        expect(searchInput).toHaveDisplayValue('');
        expect(screen.queryByTestId('clearBtn-test')).toBeNull();
        expect(component).toMatchSnapshot();
    });

    it('status error', () => {
        mockedUseAppSelector.mockReturnValue({ searchStatus: StatusModel.ERROR });
        const component = render(
            <SearchComponent />
        );
        expect(screen.getByText(/ничего не найдено/i)).toBeInTheDocument();
        expect(component).toMatchSnapshot();
    });

    it('actions calls', async() => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        mockedUseAppSelector.mockReturnValue({ searchStatus: StatusModel.IDLE });
        const component = render(
            <SearchComponent />
        );
        expect(mockedFindPostsAction).toHaveBeenCalledTimes(0);
        expect(mockedClearSearchStatusAction).toHaveBeenCalledTimes(0);

        const searchInput = screen.getByPlaceholderText(/поиск по блогу.../i);
        userEvent.type(searchInput, 'naruto');
        setTimeout(() => {
            expect(mockedFindPostsAction).toHaveBeenCalledTimes(1);
            expect(mockedFindPostsAction).toHaveBeenCalledWith({
                type: '_query', 
                payload: 'NARUTO',
            });

            setTimeout(() => {
                expect(mockedClearSearchStatusAction).toHaveBeenCalledTimes(1);
            }, 3000);
        }, 1000);
        expect(component).toMatchSnapshot();
    });
});