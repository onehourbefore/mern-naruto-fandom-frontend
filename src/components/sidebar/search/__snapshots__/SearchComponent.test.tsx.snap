// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent actions calls 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div
          class="root__search"
        >
          <div
            class="root__searchWrapper"
          >
            <img
              alt="Поиск"
              class="root__search_icon"
              src="search.png"
            />
            <input
              class="root__search_input"
              id="searchInput"
              placeholder="Поиск по блогу..."
              type="text"
              value="naruto"
            />
          </div>
          <button
            class="root__search_btn"
            data-testid="clearBtn-test"
          >
            <img
              alt="Очистить"
              class="root__search_btn_img"
              src="clear.png"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div
        class="root__search"
      >
        <div
          class="root__searchWrapper"
        >
          <img
            alt="Поиск"
            class="root__search_icon"
            src="search.png"
          />
          <input
            class="root__search_input"
            id="searchInput"
            placeholder="Поиск по блогу..."
            type="text"
            value="naruto"
          />
        </div>
        <button
          class="root__search_btn"
          data-testid="clearBtn-test"
        >
          <img
            alt="Очистить"
            class="root__search_btn_img"
            src="clear.png"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchComponent basic behavior 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div
          class="root__search"
        >
          <div
            class="root__searchWrapper"
          >
            <img
              alt="Поиск"
              class="root__search_icon"
              src="search.png"
            />
            <input
              class="root__search_input"
              id="searchInput"
              placeholder="Поиск по блогу..."
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div
        class="root__search"
      >
        <div
          class="root__searchWrapper"
        >
          <img
            alt="Поиск"
            class="root__search_icon"
            src="search.png"
          />
          <input
            class="root__search_input"
            id="searchInput"
            placeholder="Поиск по блогу..."
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchComponent status error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div
          class="root__search"
        >
          <div
            class="root__searchWrapper"
          >
            <img
              alt="Поиск"
              class="root__search_icon"
              src="search.png"
            />
            <input
              class="root__search_input"
              id="searchInput"
              placeholder="Поиск по блогу..."
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="root__notFoundStatus"
        >
          Ничего не найдено
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div
        class="root__search"
      >
        <div
          class="root__searchWrapper"
        >
          <img
            alt="Поиск"
            class="root__search_icon"
            src="search.png"
          />
          <input
            class="root__search_input"
            id="searchInput"
            placeholder="Поиск по блогу..."
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="root__notFoundStatus"
      >
        Ничего не найдено
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
